src/java.desktop/share/classes/javax/swing/plaf/metal/MetalBorders.java

```diff
@@ -275,41 +264,50 @@ public void paintBorder(Component c, Graphics g, int x, int y,
                 shadow = MetalLookAndFeel.getControlInfo();
             }
 
-            Graphics2D g2d = (Graphics2D) g;
-            AffineTransform at = g2d.getTransform();
-            Stroke oldStk = g2d.getStroke();
-            Color oldColor = g2d.getColor();
+            AffineTransform at = null;
+            Stroke oldStk = null;
+            boolean resetTransform = false;
             int stkWidth = 1;
+            double scaleFactor = 1;
 
-            // if m01 or m10 is non-zero, then there is a rotation or shear
-            // skip resetting the transform
-            boolean resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
+            if (g instanceof Graphics2D g2d) {
+                at = g2d.getTransform();
+                scaleFactor = at.getScaleX();
+                oldStk = g2d.getStroke();
+
+                // if m01 or m10 is non-zero, then there is a rotation or shear
+                // skip resetting the transform
+                resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
+
+                if (resetTransform) {
+                    g2d.setTransform(new AffineTransform());
+                    stkWidth = clipRound(Math.min(at.getScaleX(), at.getScaleY()));
+                    g2d.setStroke(new BasicStroke((float) stkWidth));
+                }
+            }
 
-            int xtranslation;
-            int ytranslation;
             int width;
             int height;
+            int xtranslation;
+            int ytranslation;
```

Comment:
I wonder why you changed the order of declaration.

Is it for consistency with the `else`-block? What about the `if` itself then?

It's very minor though.